File: binary_search/problems/search2D.js
function searchMatrix(matrix, target) {
  const m = matrix.length,
    n = matrix[0].length;
  let left = 0,
    right = m * n - 1;

  while (left <= right) {
    const mid = Math.floor(left + (right - left) / 2);
    const row = Math.floor(mid / n);
    const col = mid % n;
    const midValue = matrix[row][col];

    if (midValue === target) {
      return true;
    } else if (midValue < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  return false;
}

// Example Usage
const matrix = [
  [1, 3, 5],
  [7, 10, 11],
  [12, 16, 20],
];
console.log(searchMatrix(matrix, 10)); // Output: true

File: binary_search/problems/search2D.py
def searchMatrix(matrix, target):
  m, n = len(matrix), len(matrix[0])
  left, right = 0, m * n - 1

  while left <= right:
    mid = left + (right - left) // 2
    row, col = divmod(mid, n)
    mid_value = matrix[row][col]

    if mid_value == target:
      return True
    elif mid_value < target:
      left = mid + 1
    else:
      right = mid - 1

  return False

# Example Usage
matrix = [
  [1, 3, 5],
  [7, 10, 11],
  [12, 16, 20]
]
print(searchMatrix(matrix, 10))  # Output: True
